Code/GraphMol/MolStandardize/Fragment.cpp

```diff
@@ -14,68 +14,110 @@ typedef boost::tokenizer<boost::char_separator<char>> tokenizer;
+// constructor
+FragmentRemover::FragmentRemover() {
+  BOOST_LOG(rdInfoLog) << "Initializing FragmentRemover\n";
   FragmentCatalogParams fparams(defaultCleanupParameters.fragmentFile);
-//  unsigned int numfg = fparams->getNumFuncGroups();
-//  TEST_ASSERT(fparams->getNumFuncGroups() == 61);
+  //  unsigned int numfg = fparams->getNumFuncGroups();
+  //  TEST_ASSERT(fparams->getNumFuncGroups() == 61);
   this->d_fcat = new FragmentCatalog(&fparams);
-	this->LEAVE_LAST = true; 
+  this->LEAVE_LAST = true;
+  this->SKIP_IF_ALL_MATCH = false;
 }
 
-//overloaded constructor
-FragmentRemover::FragmentRemover(const std::string fragmentFile, const bool leave_last){
-  FragmentCatalogParams fparams(fragmentFile);
+// overloaded constructor
+FragmentRemover::FragmentRemover(const std::string fragmentFile,
+                                 bool leave_last, bool skip_if_all_match) {
+  std::string fname = !fragmentFile.empty()
+                          ? fragmentFile
+                          : defaultCleanupParameters.fragmentFile;
+  FragmentCatalogParams fparams(fname);
   this->d_fcat = new FragmentCatalog(&fparams);
-	this->LEAVE_LAST = leave_last; 
+  if (!this->d_fcat) {
+    throw ValueErrorException(
+        "could not open fragment catalog parameter file " + fname);
+  }
+  this->LEAVE_LAST = leave_last;
+  this->SKIP_IF_ALL_MATCH = skip_if_all_match;
 }
 
-//Destructor
-FragmentRemover::~FragmentRemover(){
-	delete d_fcat;
-};
+// Destructor
+FragmentRemover::~FragmentRemover() { delete d_fcat; };
 
 ROMol *FragmentRemover::remove(const ROMol &mol) {
-	BOOST_LOG(rdInfoLog) << "Running FragmentRemover\n" ;
+  BOOST_LOG(rdInfoLog) << "Running FragmentRemover\n";
   PRECONDITION(this->d_fcat, "");
   const FragmentCatalogParams *fparams = this->d_fcat->getCatalogParams();
 
   PRECONDITION(fparams, "");
 
   const std::vector<std::shared_ptr<ROMol>> &fgrps = fparams->getFuncGroups();
-  auto *removed = new ROMol(mol);
+  bool sanitizeFrags = true;
+  // provides the list of atom numbers in each fragment
+  std::vector<std::vector<int>> atomFragMapping;
+
+  // track original fragment index with the fragment itself
+  std::vector<std::pair<boost::shared_ptr<ROMol>, unsigned int>> frags;
+  for (const auto frag :
+       MolOps::getMolFrags(mol, sanitizeFrags, nullptr, &atomFragMapping)) {
+    frags.push_back(std::make_pair(frag, frags.size()));
+  }
 
   for (auto &fgci : fgrps) {
-    std::vector<boost::shared_ptr<ROMol>> frags = MolOps::getMolFrags(*removed);
-    // If nothing is left or leave_last and only one fragment, end here
-    if (removed->getNumAtoms() == 0 ||
-        (this->LEAVE_LAST && frags.size() <= 1)) {
+    size_t oCount = frags.size();
+    if (!oCount) break;
+    auto tfrags = frags;
+    // remove any fragments that this
+    frags.erase(std::remove_if(
+                    frags.begin(), frags.end(),
+                    [&fgci](const std::pair<boost::shared_ptr<ROMol>,
```

Comment:
Turns out that in c++14 we can use auto in lambda

```
 [&fgci](auto &frag) -> bool {...}
```

but not in c++11 :(

Reply:
yeah, I'm a bit nervous being too aggressive about adding c++14 features at the moment.
Maybe we can do a bit of review and consider allowing all C++14 as part of the next release cycle?